subprojects/core/src/main/java/org/gradle/internal/fingerprint/classpath/impl/ClasspathFingerprintingStrategy.java

```diff
@@ -154,6 +155,8 @@ public void visitFile(FileSystemLocationSnapshot fileSnapshot) {
                 if (normalizedContent != null) {
                     delegate.visit(fileSnapshot, normalizedContent);
                 }
+            } else if (!relativePathSegmentsTracker.isRoot()) {
+                throw new GradleException(String.format("Could not list contents of '%s'. Couldn't follow symbolic link.", fileSnapshot.getAbsolutePath()));
```

Comment:
This could also be an unavailable file, right? Maybe this would be the right place to check if the file is a symbolic link or an unavailable file? Or we leave it as you have now.

Regarding `Could not list contents of`: the snapshot is not a directory, why are you saying that we couldn't list the contents of it. Maybe `Could not read contents of `. That would be true for unreadable files and broken symlinks, right?

WDYT?

Reply:
I wanted to align it with previous behavior, i.e. https://github.com/gradle/gradle/blob/3903e8fbbc8893fe16c12709dff0f79850425b81/subprojects/snapshots/src/main/java/org/gradle/internal/snapshot/impl/DirectorySnapshotter.java#L99 and https://github.com/gradle/gradle/blob/master/subprojects/file-collections/src/main/java/org/gradle/api/internal/file/collections/DefaultDirectoryWalker.java#L111

It's not a directory there either, I'll fix the message in those cases as well, okay?

Reply:
Done